### ISSUE TITLE
[CLI] Cache GitHub API responses when running `danger pr`

### ISSUE THREAD

**Issue by macklinu** [danger/danger-js#138]
When running `danger pr <PR_URL>` often locally for testing, caching the JSON response from the GitHub API / making a [conditional request](https://developer.github.com/v3/#conditional-requests) will help prevent [exceeding the rate limit](https://developer.github.com/v3/#rate-limiting). 

See https://github.com/danger/danger/issues/198 and https://github.com/danger/danger/commit/b0b18a03fe4b0cdf0bce696b99a6dee24f84cb13 for the corresponding Danger-rb discussion/implementation. We will want to add a [`--clear-http-cache`](https://github.com/danger/danger/blob/v4.1.1/lib/danger/commands/local.rb#L15) flag for parity with Danger-rb.

If this applies to more than just `danger pr`, please chime in!

**Comment by orta** [comment]
Yeah, I think it's totally fine to store the request JSONs inside the users `tmpdir` and have the HEAD check before, can be done with a custom `fetch` on the github api object from #136 - which is how Peril works

**Comment by macklinu** [comment]
Going to work on this today. 📟

**Comment by macklinu** [comment]
I ran into a couple of things while working on this that I got caught up/confused by, but I'm going to jot down my thought process as I pick back up on this issue.

My thought on how to implement this is to use an [lru-cache](https://www.npmjs.com/package/lru-cache). When we make a GitHub API request, we can store responses in this cache with the E-Tag as a key and the JSON response as a value. We would also need to store the request URL as a key and the E-Tag as a value for looking up a previously requested resource and adding the `If-None-Match: ${etag}` header when possible.

If the response status code is 304, then we can use the cached JSON response; otherwise, proceed as normal.

We can serialize this cache and write it to the user's home directory (perhaps in `~/.cache/danger-js`) or in the [`tmpdir()`](https://nodejs.org/api/os.html#os_os_tmpdir) as you mentioned, @orta.

The `--clear-http-cache`option would just delete the serialized cache file, so the next time `danger pr` was run, a new lru-cache would be created.

I'm up for feedback on the idea. Aiming to have a PR up by the end of the week.

**Comment by orta** [comment]
Yeah - this all sounds 👍 

We'd want to to be able to have it work between separate runs ( caching in the same process is useful too, because I know I'm doing duplicated requests in the Danger setup already )

**Comment by macklinu** [comment]
Haven't made much progress on this. Figuring out how to cache the responses and update our custom `fetch()` function to support caching has been difficult. I'm curious to try out the [make-fetch-happen](https://github.com/zkat/make-fetch-happen) package, which has caching support for `fetch()`.

**Comment by orta** [comment]
It took me a while to get the pun in that name, that does looks good - some of those extra features would be nice in peril too, I'd be fine with adding this

**Comment by macklinu** [comment]
I'm wondering if this is that important if you supply a `DANGER_GITHUB_API_TOKEN` when running danger locally. Thoughts @orta?

**Comment by orta** [comment]
It's not important with the ENV var do that - but I don't think for quick one-offs people will do that
